extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java

```diff
@@ -0,0 +1,214 @@
+     * needed. The opened session is eventually closed and the marking key is removed when the provided {@link Uni} completes.
+     *
+     * @param <T>
+     * @param work
+     * @return a new {@link Uni}
+     * @see #getSession()
+     */
+    static <T> Uni<T> withLazySession(Supplier<Uni<T>> work) {
+        Context context = vertxContext();
+        if (context.getLocal(LAZY_SESSION_KEY) != null) {
+            // context already marked - no need to set the key and close the session
+            return work.get();
+        } else {
+            // mark the lazy session
+            context.putLocal(LAZY_SESSION_KEY, true);
+            // perform the work and eventually close the session and remove the key
+            return work.get().eventually(() -> {
+                context.removeLocal(LAZY_SESSION_KEY);
+                return closeSession();
+            });
+        }
+    }
+
+    /**
+     * Performs the work in the scope of a reactive transaction. An existing session is reused if possible.
+     *
+     * @param <T>
+     * @param work
+     * @return a new {@link Uni}
+     */
+    public static <T> Uni<T> withTransaction(Supplier<Uni<T>> work) {
+        return withSession(s -> {
+            return s.withTransaction(t -> work.get()
+                    .onFailure()
+                    // close the session and remove it from the vertx duplicated context
+                    .invoke(e -> closeSession()));
+        });
+    }
+
+    /**
+     * Performs the work in the scope of a reactive transaction. An existing session is reused if possible.
+     *
+     * @param <T>
+     * @param work
+     * @return a new {@link Uni}
+     */
+    public static <T> Uni<T> withTransaction(Function<Transaction, Uni<T>> work) {
+        return withSession(s -> {
+            return s.withTransaction(t -> work.apply(t)
+                    .onFailure()
+                    // close the session and remove it from the vertx duplicated context
+                    .invoke(e -> closeSession()));
+        });
+    }
+
+    /**
+     * Performs the work in the scope of a reactive session. An existing session is reused if possible.
+     *
+     * @param <T>
+     * @param work
+     * @return a new {@link Uni}
+     */
+    public static <T> Uni<T> withSession(Function<Mutiny.Session, Uni<T>> work) {
+        Context context = vertxContext();
+        Key<Mutiny.Session> key = getSessionKey();
+        Mutiny.Session current = context.getLocal(key);
+        if (current != null && current.isOpen()) {
+            // reactive session exists - reuse this session
+            return work.apply(current);
+        } else {
+            // reactive session does not exist - open a new one and close it when the returned Uni completes
+            return getSessionFactory()
+                    .openSession()
+                    .invoke(s -> context.putLocal(key, s))
```

Comment:
Doesn't the `Mutiny.SessionFactory` already associate it to the context?

Reply:
`openSession()` itself does not store the session in the context. Only `SessionFactory.withSession()` does.